### DISCUSSION TITLE
issue: `onChange` triggered inside useEffect right after initial render does not work

### DISCUSSION THREAD

**Discussion by twawszczak-wr** [react-hook-form/react-hook-form#9443] [Q&A]
### Version Number

 

&gt;= 7.20

### Codesandbox/Expo snack

https://codesandbox.io/s/dazzling-gould-yuphjz?file=/src/App.tsx:384-392

### Steps to reproduce

1. Go to sandbox page 
2. See that value for input are not recognised at form level and at field level neither

### Expected behaviour

I am trying to set value of an input inside `useEffect` right after first render*. Since version 7.20 it does not work - value is not changed, `watch` for field is not triggered. It works as expected for version 7.19.5 and lower.

* In b4 "why do u need this?" - for a select field, when there is only one option to pick, I want to choose it automatically. Possible options are fetched dynamically so typically there is a small async delay and value is set correctly. But when user is entering the same form during the same session, options are available at once. (thanks to local cache)

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by FPDK** [comment]
Hi,

When I open the sandbox in Chrome and in Firefox I see `option1` in both `Field value` and `Watched value`.

Watch is being trigged all the time, see how fast "Render count" in the top right is increasing. Try using useWatch instead:
<pre>
const selectValue = useWatch({control, name:"select"});
</pre>

**Comment by leapful** [comment]
@twawszczak-wr 

You are triggering `onChange` inside `render` prop of `Controller` component, at that time, subscribers to watch field 'select' is not registered yet thus no re-rendering happens.

A workaround with `setTimeout` when calling `onChange` will solve the issue.

Working demo: https://codesandbox.io/s/elegant-leavitt-35llt9?file=/src/Field.tsx

@FPDK 

The render count is falling to infinite loop because react-hook-form version at version 7.19.5  always creates new `onChange` function callback for `Controller` component but the `Field` component from codesandbox sample is using `useEffect` with deps `[onChange, options]` and causes `useEffect` to be performed forever (`onChange` is always changed).

Upgrade to latest version will fix the issue.
